visions/accessors.md

```diff
@@ -0,0 +1,348 @@
+
+> Note:  Accessors are also used to implement subscripts; the only real difference is that a subscript operation can have additional arguments.  This document will discuss property accesses and subscript operations interchangeably.
+
+## Two Dimensions of Access
+
+There are a wide variety of theoretically possible accessors, varying in what kind of access they provide and what mechanism they use to provide it.  Different programming languages support different combinations.  For Swift, there are three fundamental kinds of access that we are interested in:
+
+* Read access.
+* Write or “set” access.
+* Update access.
+
+Write and update access differ in that write access can create a value that did not exist before, while update access requires there be a pre-existing value.  For example, consider how the dictionary write operation `myDictionary["newKey"] = value` can create a value that did not exist before. In contrast, `myDictionary["existingKey"] += 1` modifies a value that must necessarily already exist.
+
+For each kind of access, there are multiple ways that the access might be provided.  The Swift implementation today supports three different possible mechanisms, although not all of these are currently supported by accessors:
+
+* Copying.  The value might be copied between the caller and the callee.
+* Borrowing.  The callee might provide access to the value without actually copying it.  This can be implemented by having the callee provide a pointer or reference to the value *in situ.*
+* Yielding.  Each of the above is typically provided by calling a function.  The value or pointer is passed as an argument or return value from that function.  It’s also possible for the callee to start a *coroutine* whose execution is interleaved with that of the caller.  This allows the callee to resume after the caller is finished, which is not possible for a function- or method-based accessor.
+
+## Six of Nine Fundamental Accessors
+
+Combining the possibilities above gives us a 3x3 matrix of possible accessors:
+
+|              |Read   |Write  |Update |
+|---           |---    |---    |---    |
+|**Copying**   |       |       |✗      |
+|**Borrowing** |       |✗      |       |
+|**Yielding**  |       |✗      |       |
+
+Three of these combinations turn out to be either impossible or not useful:
+
+* “Copying Update” is unnecessary.  A copying update would by definition copy the value into the caller and then copy the updated value back, which is easily achieved by using a “Copying Read” followed by a “Copying Write”
+* “Borrowing Write” and “Yielding Write” are simply impossible.  Both borrowing and yielding require an existing value.  As such, they provide “update” access.
```

Comment:
This seems to be a think-o. If you construe "copying write" (as you do below) to mean copying the source into the destination, then it is conceivable—even if not broadly useful—to borrow that source value or even to manipulate a yielded value as part of a write operation (as long as you put something back).

Here you are arguing that these are "impossible" because there is no existing value at the destination—but only in the same sense that "copying write" would be useless because you're not copying the value at the destination back to the destination.

Reply:
Remember that accessors are also used for subscript operations, and consider a collection that allocates new storage to hold a new value.  A "copying write" adds a new value to the container by copying it from the client into the container.  This works:  the write accessor accepts the value as a function argument, that function allocates the storage and puts the value into that storage.  A "borrowing write" has the container return a reference to the _existing_ value storage to the client so that the client can then update it in place without further involvement from the container.

In this scenario, such a "borrowing write" is truly impossible:  the collection doesn't even have the storage before the access begins; if it does have such storage, it may not be initialized.  This is clearest with dictionary-like containers (though Swift's standard library `Dictionary` type has a wrinkle that makes this a little tricky to demonstrate).  Such a container literally cannot yield or lend the existing value because there is no existing value to be lent or yielded.  Even a coroutine accessor (with a coroutine frame where temporary values can be stored) may not work because it may not be possible to initialize the temporary.

Does that clarify things?  (You might be thinking that by "borrowing write" I'm referring to a `set` accessor that accepts its argument as a borrow.  That is _not_ what I am attempting to describe here.  I'll play with the explanation to see if I can clarify it.)

Reply:
Yes, I understand: My feedback is that the terminology is used inconsistently and the 3x3 matrix is hence not a systematic grid—as you say, you use the term "copying write" to mean a copy from the client to the container, but you _don't_ use "borrowing write" to mean a borrow from the client to the container.

If the directionality were consistently aligned with what you mean by "borrowing write," then a "copying write" (and hence the entire "write" column) would be impossible and/or useless by construction, since there's likewise no way to copy a non-existent value from the container.